feat: Implement plant dashboard

GoogolGenius opened the pull request:
# feat: Plant Dashboard & Device Management

## Overview
This PR implements a plant monitoring dashboard with CRUD operations for plant monitors (connected ESP devices) featuring a responsive UI utilizing Tailwind CSS, as well as graphs for data visualization powered by the chart.js lib.


## Features

### Plant Registration & Management
- Implemented a complete serial registration system on the web app, allowing for physical ESP hardware to connect
- Detects physical hardware devices connected through USB (serial) on macOS to send appropriate registration info
- Full Create, Read, Update, Delete routes via backend routes on the express.js server for plant monitoring setup
- Frontend modal UI for plant/device onboarding

### ESP Device EEPROM Clear Support
- On serial startup, if a `CLEAR` command is detected within 5 secs, the EEPROM is wiped and the ESP8266 device enters configuration mode
- Optional “Clear device EEPROM” feature available when removing a plant, triggering new API endpoint for wiping the storage and resetting the configuration as needed

### Dashboard & Detail Views
- **Main Dashboard**: Plant cards are dynamically loaded; they display the latest moisture, name, location, and last-updated timestamp in an intuitive overview
- **Detail Page**: Uses Chart.js library to display plant moisture trend over time after pulling from the SQL db

### Improved UI/UX
- Responsive, modern design with Tailwind CSS with enhancements for modals, cards, buttons, and loading/error states
- Bio::Neos Company Branding added
- Mobile optimizations for viewability on narrow-width screens
- Automatic dark/light mode depending on browser settings


## Future Direction

- [x] Implement **Remove Plant** UI option
- [ ] Implement plant configuration editing
- [ ] Add support for Windows devices during config
- [ ] Add documentation: update README with dashboard screenshots, API docs
- [ ] Add tests for API endpoints and frontend components


## Images
Main plant dashboard UI:

<img width="1800" height="1130" alt="image" src="https://github.com/user-attachments/assets/d2ea5f0f-ca02-40cc-9359-0ed3e6cf093f" />

<img width="1800" height="1130" alt="image" src="https://github.com/user-attachments/assets/aae8d0f3-e7e1-47fc-bd41-a567216104f6" />


Detailed individual plant info UI:

<img width="1800" height="1130" alt="image" src="https://github.com/user-attachments/assets/1b77e085-ab17-44b6-886d-27fddb0edfb3" />

<img width="1800" height="1130" alt="image" src="https://github.com/user-attachments/assets/74e3474e-1956-4df0-a46d-eb8298678ebe" />


Plant device register workflow UI:

<img width="1800" height="1130" alt="image" src="https://github.com/user-attachments/assets/532f80d1-65a8-42fe-a1e2-6c8bc9709d88" />

<img width="1800" height="1130" alt="image" src="https://github.com/user-attachments/assets/9f5f31b0-e459-423e-b76d-17a87e106726" />


Plant delete workflow UI:

<img width="1800" height="1130" alt="image" src="https://github.com/user-attachments/assets/79913fa0-22f7-4f6f-bc8b-67f8609159b5" />

<img width="1800" height="1130" alt="image" src="https://github.com/user-attachments/assets/60ab1366-9416-404c-a5fa-abad26cd3df5" />

GoogolGenius committed:
Add PUT and DELETE routes for plants

GoogolGenius committed:
Add error msg for no plant found in saturation method

GoogolGenius committed:
Add name and location fields to esp module config

Extended the ModuleConfig struct to include 'name' and 'location' fields. Updated configuration parsing, validation, and POST request body to handle these new fields, allowing each device to report its name and location along with sensor data

GoogolGenius committed:
Implement plant dashboard card UI

GoogolGenius committed:
Add Last Updated indicator

GoogolGenius committed:
Implement additional detailed plant page

GoogolGenius committed:
Update styling

GoogolGenius committed:
Revert "Add name and location fields to esp module config"

This reverts commit 7bfc6676eadc95670a95a7235a6eb568becff0b4.

GoogolGenius committed:
Add EEPROM clear option via serial on startup

On startup, the device now waits 5 seconds for a 'CLEAR' command over serial. If received, it wipes the EEPROM and enters configuration mode, allowing easier reset of stored connection info without reflashing.

GoogolGenius committed:
Complete CRUD modal impl for plant device registration

- Implement full device configuration workflow with ESP8266 serial communication
- Add better plant database registration with duplicate MAC validation
- Fix moisture level display bug (was showing 100% for new plants)
- Add better error handling and user feedback

GoogolGenius committed:
Update output.css

GoogolGenius committed:
Add plant deletion with optional device EEPROM clear

Introduces UI and backend support for removing a plant from the database, including a confirmation modal and the option to clear the associated device's EEPROM via a new API endpoint. Updates CSS for new modal and button styles, and enhances the frontend logic to handle device selection, confirmation, and error handling during the deletion process.

GoogolGenius committed:
Add favicon to web app

Added favicon.png to the public/img directory and updated index.html and plant-detail.html to include the favicon link in the head section

GoogolGenius committed:
Align header title to the right

Changed the header title alignment from center to right in the main page to improve layout on narrow screens

GoogolGenius commented:
One thing @peterz0310 brought up is: how will the MAC address be obtained by the user? The form requires it because we want to register the plant in the database with the appropriate MAC address. Could we establish a 2-way serial communication? Or is there a better impl?

sgdavis1 commented:
There is an ESP API call to get it. https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/network/esp_wifi.html#_CPPv416esp_wifi_get_mac16wifi_interface_tAL6E_7uint8_t

sgdavis1 commented:
Also we shouldn't have to ask the user what it is (as this would introduce the possibility of human error). We can update from here but that will be important. Actually "showing" it to the user is a good way for them to identify that the serial connection was completed and worked. So I guess the big missing piece was the ability to send a command to the ESP to query it for the interface MAC. We need to get this implemented in some way but I want to think through the workflow first. Thanks for getting things this far! It is close!

sgdavis1 commented:
Or rather maybe this: https://arduino-esp8266.readthedocs.io/en/latest/esp8266wifi/station-class.html#macaddress

GoogolGenius committed:
Automate MAC address detection during registration

The ESP8266 firmware now sends its MAC address on startup and during configuration mode. The web app and API have been updated to automatically detect the MAC address from the device, removing the need for manual entry. Serial communication uses line parsing for more robust response handling, and the plant registration flow now uses the detected MAC address for database association

GoogolGenius reviewed (commented):
This might work, idk... i just edited the esp code to send the mac address automatically

GoogolGenius commented:
I already see a problem with this though, if you connected the esp but didn't have the web server and app with the menu running, it wouldn't be able to grab the mac address being sent at that time. i'll fix that some other time

GoogolGenius committed:
Address bug with MAC address receive/response

I don't know why tailwind is adding new CSS...

GoogolGenius commented:
I fixed it by sending the MAC address from the esp after the configuration was received from the web app, but the web app will not proceed until it receives it correctly from the esp.

GoogolGenius commented:
Good luck reviewing all of this btw, haha. It's a ton, but I had a lot of fun working on this project during my internship at Bio::Neos! I hope it can be merged soon! Thank you again for all the support and for providing me with this opportunity.
